docs/src/docs/asciidoc/maven-plugin.adoc

```diff
@@ -509,43 +509,35 @@ experimental features such as `experimental-class-loader-support` or advanced fe
-    - To configure options for your tests, use the name `test`.
-    - To configure additional sets of options, declare each with a unique name other than
-      `main` or `test`.
-* The `main` options are enabled automatically whenever your application is run with the
-  agent.
-* The `test` options are enabled automatically whenever your tests are run with the agent.
-* To enable any other set of named `<options>`, supply `-DagentOptions=<NAME>` as a
-  command-line argument for Maven, where `<NAME>` corresponds to the `name` attribute of
-  the `<options>` element.
-
-[WARNING]
-====
-The Native Image Maven plugin automatically configures the `config-output-dir` for the
-agent. An attempt to configure a custom value for the `config-output-dir` option will
-therefore result in a build failure.
-====
+- you can specify them in your POM as described bellow.
 
-The following example is likely more complex than anything you would do in your own
-projects, but it demonstrates how to configure four sets of `<options>`.
+[source,xml,indent=0]
+include::../snippets/maven/pom.xml[tag=native-plugin-agent-configuration]
 
-* The unnamed set is always active.
-* The `main` set is automatically active for application execution.
-* The `test` set is automatically active for test execution.
-* The `periodic-config` set is never active by default, but it can be enabled via
-  `-DagentOptions=periodic-config` on the command line.
+This example shows all possibilities you can use for native agent configuration. If you don't need some of the options,
+just remove them.
+
+Agent can be run in one of the following modes:
+
+* `standard` - in this mode you run agent only with options provided in `commonOptions` section.
+* `direct` - in this mode you can provide command line that you want to be executed as an agent configuration. In this
+     mode, user is fully responsible for agent configuration, and the rest of the config file will be ignored.
+* `conditional` - in this mode you can provide additional files that can be used as a filter for the agent output (https://github.com/oracle/graalvm-reachability-metadata/blob/master/docs/CollectingMetadata.md[see this]). You
+can read more about conditional mode https://www.graalvm.org/22.1/reference-manual/native-image/ExperimentalAgentOptions/#generating-conditional-configuration-using-the-agent[here].
+
+Each option and how you should use it is described bellow:
+
+* `enabled` - a simple flag that specifies whether the agent is on or off. Can be set to true or false.
+* `defaultMode` - agent mode switcher. Can be set to: standard, direct or conditional.
+* `modes` - list of additional mode options, specific for the mode type. Inside this tag, you can specify options for
+ direct or conditional modes. Standard mode doesn't have any specific options.
+* `commonOptions` - list of options that can be specify independent from agent mode. More about common options can be found
```

Comment:
Could you elaborate on the choice of `commonOptions` versus `options`?

Reply:
First reason for this choice is to align with `gradle` implementation and have uniform agent manipulation mechanism. Also, with `options` there are no semantic meaning for any option (you can put whatever you want inside '<option>` tag). On the other hand, with `commonOptions` we have named options, which makes the file more readable, and easier for parsing.

Reply:
> First reason for this choice is to align with gradle implementation and have uniform agent manipulation mechanism.

Not sure what you mean about Gradle implementation, but I see no usage of `commonOptions` in https://graalvm.github.io/native-build-tools/latest/gradle-plugin.html.

> Also, with options there are no semantic meaning for any option (you can put whatever you want inside <option> tag). On the other hand, with commonOptions we have named options, which makes the file more readable, and easier for parsing.

What prevents us to use an `<options>` element with named options?

Reply:
I think what David is trying to say is better understood with an example:

```xml                              
<modes>
	<conditional>
		<userCodeFilterPath>user-code-filter.json</userCodeFilterPath>
		<parallel>true</parallel>
	</conditional>
</modes>
<commonOptions>
	<builtinCallerFilter>true</builtinCallerFilter>
	<builtinHeuristicFilter>true</builtinHeuristicFilter>
	<enableExperimentalPredefinedClasses>false</enableExperimentalPredefinedClasses>
	<enableExperimentalUnsafeAllocationTracing>false</enableExperimentalUnsafeAllocationTracing>
	<trackReflectionMetadata>true</trackReflectionMetadata>
</commonOptions>
```

`modes.conditional` is mapped with a `java.util.Properties`, therefore, children tags can be whatever. IIRC, they are passed as-is to GraalVM.

`commonOptions`, on the other hand, is mapped with `CommonOptionsConfiguration` class, which has a well-know set of options mapped as `@Parameters`.

Having said that, I think all Sébastien is asking (please correct me if I'm wrong) is to simply rename it as `options`, something I would agree with.

Reply:
Yep